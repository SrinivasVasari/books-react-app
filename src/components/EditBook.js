import React, { useState, useContext } from "react";
import BooksContext from "../context/context";

const EditBook = () => {
  const { state, dispatch } = useContext(BooksContext);
  const [bookName, setBookName] = useState(
    state && state.currentBook.books.bookName
  );
  const [bookAuthor, setBookAuthor] = useState(
    state && state.currentBook.books.bookAuthor
  );
  const [bookDesc, setBookDesc] = useState(
    state && state.currentBook.books.bookDesc
  );

  const handleSubmit = (event) => {
    event.preventDefault();
    const updateBook = {
      bookName,
      bookAuthor,
      bookDesc,
    };
    if (
      bookName.trim() === "" ||
      bookAuthor.trim() === "" ||
      bookDesc.trim() === ""
    ) {
      alert("Please enter the book details");
    } else {
      dispatch({ type: "UPDATE_BOOK", payload: updateBook });
      setBookName("");
      setBookAuthor("");
      setBookDesc("");
    }
  };

  return (
    <div className="container main-content">
      <div className="editBookForm">
        <h4 className="center-align">Update Book</h4>
        <form className="form col s12" onSubmit={handleSubmit}>
          <div className="row">
            <div className="input-field col s12">
              <input
                type="text"
                name="name"
                id="name"
                placeholder="Book Name"
                onChange={(e) => setBookName(e.target.value)}
                value={bookName || ""}
                className="name"
              />
            </div>
            <div className="input-field col s12">
              <input
                type="text"
                name="author"
                id="author"
                placeholder="Book Author"
                onChange={(e) => setBookAuthor(e.target.value)}
                value={bookAuthor || ""}
                className="Author"
              />
            </div>
          </div>
          <div className="row">
            <div className="input-field col s12">
              <input
                type="text"
                onChange={(e) => setBookDesc(e.target.value)}
                value={bookDesc || ""}
                name="description"
                id="description"
                placeholder="Book Description"
                className="description"
              />
            </div>
          </div>
          <button type="submit" className="waves-effect waves-light btn-small">
            update Book
          </button>
        </form>
      </div>
    </div>
  );
};

export default EditBook;
