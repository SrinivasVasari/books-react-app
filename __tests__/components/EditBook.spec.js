import React from "react";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";
import EditBook from "../../src/components/EditBook";
import BooksContext from "../../src/context/context";

describe("Edit Book Container", () => {
  let wrapper;
  let alertSpy;
  beforeEach(() => {
    alertSpy = jest.spyOn(window, "alert");
    wrapper = shallow(<EditBook />);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("Should render correctly", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it("Should render form currentBook is null", () => {
    const state = {
      books: [
        {
          books: {
            bookAuthor: "author",
            bookDesc: "desc",
            bookName: "name,",
          },
          id: "1",
        },
      ],
      currentBook: {
        books: {
          bookAuthor: "author",
          bookDesc: "desc",
          bookName: "name,",
        },
        id: "1",
      },
    };
    const wrapper = mount(
      <BooksContext.Provider value={{ state }}>
        <EditBook {...state} />
      </BooksContext.Provider>
    );
    expect(wrapper).not.toBeNull();
  });
  it("Should call handleSubmit function when form is submitted", () => {
    const dispatchMock = jest.fn();
    const handleSubmitFn = jest.fn();
    const fakeEvent = { preventDefault: () => {} };
    const wrapper = mount(
      <BooksContext.Provider value={{ dispatch: dispatchMock }}>
        <EditBook onSubmit={handleSubmitFn} />
      </BooksContext.Provider>
    );
    const form = wrapper.find("form");
    wrapper.find("#name").simulate("change", {
      target: { name: "name", value: "blah" },
    });
    wrapper.find("#author").simulate("change", {
      target: { name: "author", value: "blah1" },
    });
    wrapper.find("#description").simulate("change", {
      target: { name: "description", value: "blah2" },
    });
    form.simulate("submit", fakeEvent);
    expect(handleSubmitFn).toHaveBeenCalledTimes(0);
  });
  it("Should call alert inside handleSubmit function when when submitted with empty values", () => {
    const dispatchMock = jest.fn();
    const handleSubmitFn = jest.fn();
    const fakeEvent = { preventDefault: () => {} };
    const domAlert = window.alert;
    window.alert = () => {};
    const wrapper = mount(
      <BooksContext.Provider value={{ dispatch: dispatchMock }}>
        <EditBook onSubmit={handleSubmitFn} />
      </BooksContext.Provider>
    );
    const form = wrapper.find("form");
    wrapper.find("#name").simulate("change", {
      target: { name: "name", value: "" },
    });
    form.simulate("submit", fakeEvent);
    expect(handleSubmitFn.mock.calls).not.toBeNull();
    expect(alertSpy).not.toBeNull();
    window.alert = domAlert;
  });
});
