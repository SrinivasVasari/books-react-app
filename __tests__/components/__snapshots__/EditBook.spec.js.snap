// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit Book Container Should render correctly 1`] = `
<div
  className="container main-content"
>
  <div
    className="editBookForm"
  >
    <h4
      className="center-align"
    >
      Update Book
    </h4>
    <form
      className="form col s12"
      onSubmit={[Function]}
    >
      <div
        className="row"
      >
        <div
          className="input-field col s12"
        >
          <input
            className="name"
            id="name"
            name="name"
            onChange={[Function]}
            placeholder="Book Name"
            type="text"
            value=""
          />
        </div>
        <div
          className="input-field col s12"
        >
          <input
            className="Author"
            id="author"
            name="author"
            onChange={[Function]}
            placeholder="Book Author"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="input-field col s12"
        >
          <input
            className="description"
            id="description"
            name="description"
            onChange={[Function]}
            placeholder="Book Description"
            type="text"
            value=""
          />
        </div>
      </div>
      <button
        className="waves-effect waves-light btn-small"
        type="submit"
      >
        update Book
      </button>
    </form>
  </div>
</div>
`;
